Sources/NIOHPACK/HeaderTables.swift

```diff
@@ -0,0 +1,310 @@
+/// avoid allocation churn while evicting and replacing entries.
+struct HeaderTableStorage {
+    static let defaultMaxSize = 4096
+    
+    private var buffer: StringRing
+    private var headers: CircularBuffer<HeaderTableEntry>
+    
+    private(set) var maxSize: Int
+    private(set) var length: Int = 0
+    
+    var count: Int {
+        return self.headers.count
+    }
+    
+    init(allocator: ByteBufferAllocator, maxSize: Int = HeaderTableStorage.defaultMaxSize) {
+        self.maxSize = maxSize
+        self.buffer = StringRing(allocator: allocator, capacity: self.maxSize)
+        self.headers = CircularBuffer(initialRingCapacity: self.maxSize / 64)    // rough guess: 64 bytes per header
+    }
+    
+    init(allocator: ByteBufferAllocator, staticHeaderList: [(String, String)]) {
+        // calculate likely total byte length we will actually need
+        // the static table is all ASCII, so character count == byte count here,
+        // so we don't need to convert to UTF-8 bytes just to get byte length.
+        let bytesNeeded = staticHeaderList.reduce(0) { $0 + $1.0.count + $1.1.count }
+        
+        // now allocate & encode all the things
+        self.buffer = StringRing(allocator: allocator, capacity: bytesNeeded)
+        self.headers = CircularBuffer(initialRingCapacity: staticHeaderList.count)
+        
+        var len = 0
+        
+        for (name, value) in staticHeaderList {
+            let nameStart = self.buffer.ringTail
+            let nameLen = try! self.buffer.write(string: name)
+            let valueStart = self.buffer.ringTail
+            let valueLen = try! self.buffer.write(string: value)
+            
+            let entry = HeaderTableEntry(name: HPACKHeaderIndex(start: nameStart, length: nameLen),
+                                         value: HPACKHeaderIndex(start: valueStart, length: valueLen))
+            
+            // the input list is already in order, so we push at the end of our internal list
+            self.headers.append(entry)
+            len += entry.length
+        }
+        
+        self.length = len
+        self.maxSize = len
+    }
+    
+    subscript(index: Int) -> HeaderTableEntry {
+        return self.headers[index]
+    }
+    
+    subscript(name: String) -> [String] {
+        return self.findHeaders(matching: name.utf8).map { self.string(idx: $0.value) }
+    }
+    
+    @_specialize(where C == String.UTF8View)
+    func findHeaders<C: Collection>(matching name: C) -> [HeaderTableEntry] where C.Element == UInt8 {
+        return self.headers.lazy.filter {
+            self.buffer.equalCaseInsensitiveASCII(view: name, at: $0.name)
+        }
+    }
+    
+    @_specialize(where C == String.UTF8View)   // from String-based API
+    @_specialize(where C == ByteBufferView)    // from HPACKHeaders-based API
+    func indices<C: Collection>(matching name: C) -> [Int] where C.Element == UInt8 {
+        // Returns a LazyFilterCollection
```

Comment:
The compiler considers it to be a `LazyMapSequence<LazyFilterSequence<EnumeratedSequence<CircularBuffer<HeaderTableEntry>>>, Int>`. Looks like the compiler starts complaining only once I use a `let result = ...; return result`. I could've sworn there was something in the Swift manifesto about never intrinsically converting between types… 😠

Since it's internal I'll just modify the return type. Seems weird though, when the return type is longer than the method name itself…

Reply:
…and also exposes in the type name a ton of useless information concerning the internals of the object. Sigh.

Reply:
Hrm. `AnyCollection` and `AnySequence` have some undesirable performance properties that we probably want to avoid on a path this hot.

It may be that the best way to handle this is to either introduce a wrapper type or a typealias for this purpose. The long name of the type isn't really a big deal, except for typing it and having to see it, so the typealias would address that concern. Alternatively, we can concretise this by writing a wrapper struct that stores the actual type in a private member variable.

How do those ideas seem to you?